neuralmonkey/decoders/beam_search_decoder.py

```diff
@@ -62,127 +72,190 @@ def beam_size(self):
-        next_beam_prev_attns = [tf.gather(a, next_beam_ids) for a in attns]
-        next_lengths = tf.gather(hyp_lengths, next_beam_ids)
-
-        # update finished flags
-        has_just_finished = tf.equal(next_word_ids, END_TOKEN_INDEX)
-        next_finished = tf.logical_or(
-            tf.gather(bs_state.finished, next_beam_ids),
-            has_just_finished)
-
-        output = SearchStepOutput(
-            scores=topk_scores,
-            parent_ids=next_beam_ids,
-            token_ids=next_word_ids)
-
-        search_state = SearchState(
-            logprob_sum=next_logprob_sum,
-            lengths=next_lengths,
-            finished=next_finished,
-            last_word_ids=next_word_ids,
-            last_state=next_beam_prev_state,
-            last_attns=next_beam_prev_attns)
-
-        return search_state, output
-    # pylint: enable=too-many-locals
+        def cond(*args) -> tf.Tensor:
+            bsls = BeamSearchLoopState(*args)
+            return tf.less(bsls.decoder_loop_state.step, self._max_steps)
+
+        final_state = tf.while_loop(
+            cond,
+            self.get_body(att_objects),
+            self.get_initial_loop_state(att_objects))
+
+        scores = final_state.bs_output.scores.stack()
+        parent_ids = final_state.bs_output.parent_ids.stack()
+        token_ids = final_state.bs_output.token_ids.stack()
+
+        return SearchStepOutput(scores=scores,
+                                parent_ids=parent_ids,
+                                token_ids=token_ids)
+
+    def get_body(self, att_objects: List[BaseAttention]) -> Callable[
+            [BeamSearchLoopState], BeamSearchLoopState]:
+
+        decoder_body = self.parent_decoder.get_body(att_objects, False)
+
+        # pylint: disable=too-many-locals
+        def body(*args) -> BeamSearchLoopState:
+            loop_state = BeamSearchLoopState(*args)
+
+            bs_state = loop_state.bs_state
+            dec_loop_state = loop_state.decoder_loop_state
+
+            # recreate loop state for the decoder body function
+            input_dec_loop_state = LoopState(
+                step=dec_loop_state.step,
+                input_symbol=bs_state.last_word_ids,
+                train_inputs=self.parent_decoder.train_inputs,
+                prev_rnn_state=bs_state.last_state,
+                # TODO put something else here to work with LSTM:
+                prev_rnn_output=bs_state.last_state,
+                # TODO put something else here:
+                rnn_outputs=dec_loop_state.rnn_outputs,
+                # TODO put something else here:
+                prev_logits=dec_loop_state.prev_logits,
+                # TODO put something else here:
+                logits=dec_loop_state.logits,
+                prev_contexts=bs_state.last_attns,
+                # TODO put something else here:
+                mask=dec_loop_state.mask,
+                # TODO put something else here:
```

Comment:
protože když jsem to psal, tak jsem neměl rozmyšlený, co tam patří a co ne. Vůbec si timhle kódem vlastně nejsem jistej.